### DISCUSSION TITLE
Time stamp with serial module

### DISCUSSION THREAD

**Discussion by WoodenNautilus** [Eugeny/tabby#6096] [General]
Hello,

First of all, I would like to thank you for the great work you did with this software !

I have a question. Do you plan on adding an option to add a timestamp when a new line appears on the serial link ? It would be a great addition !

Thanks for your answer !

Greetings

**Comment by Eugeny** [comment]
This is something that a plugin could do easily by providing a [SessionMiddleware](https://docs.tabby.sh/terminal/classes/SessionMiddleware.html) :)

**Reply by WoodenNautilus** [reply]
I am completely out of my confort zone when it comes to TypeScript. Thanks for the info, I will try to have a look someday !

**Reply by WoodenNautilus** [reply]
Hello Eugeny,

Would you have some examples or insights to provide on how to provide a SessionManager ? This would probably the hardest part for me.

**Reply by Eugeny** [reply]
Sure, check out how this middleware parses and changes session output: https://github.com/Eugeny/tabby/blob/master/tabby-terminal/src/middleware/oscProcessing.ts

**Reply by WoodenNautilus** [reply]
Thanks a lot ! Another question, what Plugin skeleton can I use as a base example ? Is the Clippy one suitable or is there another one more appropriate ?

**Reply by Eugeny** [reply]
Yes, you can use the clippy plugin as a template.

**Reply by williamblair333** [reply]
Would you be willing to take a kickstarter like / bounty like  and implement timestamping per line into tabby?

**Reply by miniol** [reply]
@WoodenNautilus, were you be able to write the plugin? I'm interested in timestamps too.

**Reply by WoodenNautilus** [reply]
Hello,

I haven't been able to develop this plugin. Though, I will have time in the incoming weeks and hope to be able to develop this feature. I will let you know I soon as I start this project,

**Reply by Eugeny** [reply]
@WoodenNautilus let me know if you'll have questions. The cleanest way to add timestamps to the output would be implementing a [SessionMiddleware](https://docs.tabby.sh/terminal/classes/SessionMiddleware.html) which can read and change the output

**Reply by WoodenNautilus** [reply]
Hello guys,

I wanted to let you know that I just released the first version of this plugin called serial-timestamp. It is available under the Plugins menu of Tabby.

Let me know on the [Github repos](https://github.com/WoodenNautilus/tabby-serial-timestamp) of this plugin if you have any issue.

A huge thank you to @Eugeny. Without him I wouldn't have been able to create this plugin.

**Reply by Eugeny** [reply]
tested and works nicely, great job 👌

**Reply by WoodenNautilus** [reply]
Thanks @Eugeny. Thank you for helping me, for developing Tabby and for making it an easily customizable tool !

**Comment by WoodenNautilus** [comment]
Hello @Eugeny ,

I have started this project. Until now I have been able to create a settings tab to configure the serial timestamp (none, dd/mm/yyyy and mm/dd/yyyy). 

~~On the settings tab of the Clippy plugin, I have a question. I do not understand where the "paperclip" icon declared inside the ClippySettingsTabProvider class is located. How does the program knows where to find it ?~~ After digging some more, I found icons.json

I am looking into the documentation and examples of the SessionMiddleware (and the class extending it TerminalStreamProcessor and oscProcessor). I am struggling to understand how I would "connect" to tabby-serial once I will be done writing my SessionMiddleware. 
Until now, I failed to import SessionMiddleware from the tabby-core api.
I tried to declare a prodiver in my index.ts but I am not sur this is how it should be done.
You can a look a the code [here](https://github.com/WoodenNautilus/tabby-serial-timestamp).
If you could point me to some examples or give me some explanations, it would be much appreciated !

I hope my questions are not dumb questions. If so, please forgive me, I don't know much about Web technologies and Electron applications.

Thanks for your help.

**Reply by Eugeny** [reply]
Hey thanks for picking this up! To tell tabby-serial to use your SessionMiddleware, you can just add it to the active session in your TerminalDecorator here: https://github.com/WoodenNautilus/tabby-serial-timestamp/blob/master/src/terminalDecorator.ts#L30

Something along these lines should do the trick:

```typescript
class SerialTimestampMiddleware extends SessionMiddleware {
	//...
    feedFromSession (data: Buffer): void {
        for (const line of data.toString().split('\n')) {
            const formatted = `[${new Date().toLocaleTimeString()}] ${line}\n`
            this.outputToTerminal.next(Buffer.from(formatted))
        }
    }
}

// in TerminalDecorator.attachToSession:
session.middleware.push(new SerialTimestampMiddleware())
```

**Reply by WoodenNautilus** [reply]
Thanks for the valuable input.

I have tried to implement the hints given above.
It is not 100% identical to what you wrote because I tried to replicate what is done in class **ZModemDecorator** of tabby-terminal.

Unfortunately, it is looking like the **middleware** attribute is not exposed to a plugin. I do see it declared inside **session.ts** but it does not appear inside **session.d.ts** which is used by the Intellisense. When compiling I get errors about **BaseSession** not having a property **middleware**.

In addition, I cannot import **SessionMiddleware** from **tabby-terminal** in my SerialTimestampSessionMiddleware. It is probably a stupid mistake but I cannot figure it out.

The code has been updated on my GitHub to reflect your suggestions.

I feel a little bit guilty because I am absolutely not autonomous on this dev. Thus let me thank you for helping me.

**Reply by Eugeny** [reply]
No worries! The clippy plugin is referencing an older version of the tabby plugins - just update all the `tabby-*` packages in your package.json to their latest versions.

**Reply by WoodenNautilus** [reply]
I feel one more time stupid as I looked into this and failed to understand properly the syntax **_^1.0.147_**. I thought it meant npm would fetch the latest version available higher or equal to 1.0.147 and lower than 2.0.0.

Thanks to your help, I am now able to add a dummy message at the beginning of each line on a terminal.
My next step will be to actually display a timestamp in the format provided by the settings.

I tried to found a way to attach to the session if it is a serial session only but could not find a way to do it. Would you have some hints on how to do this ?

**Reply by Eugeny** [reply]
Yes, you can check whether `terminal instanceof SerialTabComponent` in your TerminalDecorator

**Reply by WoodenNautilus** [reply]
I just tried and face a problem. SerialTabComponent is not exported by tabby modules. How can I reference it in my plugin ?

**Reply by Eugeny** [reply]
True - I've added the exports to master, but to make the plugin work with the existing Tabby versions, I suggest using a different approach at least in the beginning - try `terminal.constructor.name === 'SerialTabComponent'`

**Reply by WoodenNautilus** [reply]
Thanks a lot, it works like a charm !
When the new release will roll out, I will make the code evolve to use the more elegant methods suggested earlier.

**Comment by WoodenNautilus** [comment]
As this plugin seems to work, I am closing this discussion. Thanks again for your help.
